plugins/docker/src/main/java/org/radargun/service/DockerService.java

```diff
@@ -185,23 +223,85 @@ private void printContainerLog(String containerId) {
          .exec(loggingCallback);
 
       try {
-         loggingCallback.awaitCompletion(3, TimeUnit.SECONDS);
+         if (!logLatch.await(awaitLogTimeout, TimeUnit.MILLISECONDS)) {
+            throw new RuntimeException("Timed out reading log");
+         }
       } catch (InterruptedException e) {
-         log.error(e.getMessage());
          Thread.currentThread().interrupt();
+         throw new RuntimeException("Interrupted reading log", e);
       }
    }
 
    private class LogCallback extends LogContainerResultCallback {
+      private CountDownLatch latch;
+      private final Pattern pattern;
+      private ByteBuffer buffer = ByteBuffer.allocate(1024);
+
+      public LogCallback(CountDownLatch latch, Pattern pattern) {
+         this.latch = latch;
+         this.pattern = pattern;
+      }
+
       @Override
       public void onNext(Frame frame) {
-         log.info(frame.toString());
+         byte[] payload = frame.getPayload();
+         // Remove color escape sequences
+         if (buffer.capacity() < payload.length) {
+            buffer = ByteBuffer.allocate(Math.max(payload.length, buffer.capacity() * 2));
+         } else {
+            buffer.rewind();
+         }
+         for (int i = 0; i < payload.length; i++) {
+            byte b = payload[i];
+            if (b == 27) { // ESCAPE
+               if (++i < payload.length && (b = payload[i]) == '[') {
+                  while (++i < payload.length) {
+                     b = payload[i];
+                     if (b >= '@' && b <= '~') break;
```

Comment:
What does "color escape sequence" look like? This break statement is called even if there's [ or ] so I'm trying to see what's the goal/purpose.

Reply:
It is called only when the `[` is preceded by an escape char. Subsequently there are some numbers and the sequence is terminated with a letter (included).